extensions-core/namespace-lookup/src/main/java/io/druid/query/lookup/NamespaceLookupExtractorFactory.java

```diff
@@ -0,0 +1,297 @@
+  private final boolean injective;
+
+  private final String extractorID;
+
+  @JsonCreator
+  public NamespaceLookupExtractorFactory(
+      @JsonProperty("extractionNamespace") ExtractionNamespace extractionNamespace,
+      @JsonProperty("firstCacheTimeout") Long firstCacheTimeout,
+      @JsonProperty("injective") boolean injective,
+      @JacksonInject final NamespaceExtractionCacheManager manager
+  )
+  {
+    this.extractionNamespace = Preconditions.checkNotNull(
+        extractionNamespace,
+        "extractionNamespace should be specified"
+    );
+    this.firstCacheTimeout = firstCacheTimeout == null ? DEFAULT_SCHEDULE_TIMEOUT : firstCacheTimeout;
+    Preconditions.checkArgument(this.firstCacheTimeout >= 0);
+    this.injective = injective;
+    this.manager = manager;
+    this.extractorID = buildID();
+    this.lookupIntrospectHandler = new LookupIntrospectHandler()
+    {
+      @GET
+      @Path("/keys")
+      @Produces(MediaType.APPLICATION_JSON)
+      public Response getKeys()
+      {
+        try {
+          return Response.ok(getLatest().keySet()).build();
+        }
+        catch (ISE e) {
+          return Response.status(Response.Status.NOT_FOUND).entity(ServletResourceUtils.sanitizeException(e)).build();
+        }
+      }
+
+      @GET
+      @Path("/values")
+      @Produces(MediaType.APPLICATION_JSON)
+      public Response getValues()
+      {
+        try {
+          return Response.ok(getLatest().values()).build();
+        }
+        catch (ISE e) {
+          return Response.status(Response.Status.NOT_FOUND).entity(ServletResourceUtils.sanitizeException(e)).build();
+        }
+      }
+
+      @GET
+      @Path("/version")
+      @Produces(MediaType.APPLICATION_JSON)
+      public Response getVersion()
+      {
+        final String version = manager.getVersion(extractorID);
+        if (null == version) {
+          // Handle race between delete and this method being called
+          return Response.status(Response.Status.NOT_FOUND).build();
+        } else {
+          return Response.ok(ImmutableMap.of("version", version)).build();
+        }
+      }
+
+      @GET
+      @Produces(MediaType.APPLICATION_JSON)
+      public Response getMap()
+      {
+        try {
+          return Response.ok(getLatest()).build();
+        }
+        catch (ISE e) {
+          return Response.status(Response.Status.NOT_FOUND).entity(ServletResourceUtils.sanitizeException(e)).build();
+        }
+      }
+
+      private Map<String, String> getLatest()
+      {
+        return ((MapLookupExtractor) get()).getMap();
+      }
+    };
+  }
+
+  @VisibleForTesting
+  public NamespaceLookupExtractorFactory(
+      ExtractionNamespace extractionNamespace,
+      NamespaceExtractionCacheManager manager
+  )
+  {
+    this(extractionNamespace, null, false, manager);
+  }
+
+  @Override
+  public boolean start()
+  {
+    final Lock writeLock = startStopSync.writeLock();
+    writeLock.lock();
```

Comment:
is there really a benefit to using a read-write lock here? was a simple synchronized block not fast enough?